### PR TITLE
compare toString on SafeArg

### PR DESCRIPTION
**Pull request by pnepywoda**

### CODE REVIEW

File: test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java [hunk 1]
```diff
@@ -41,8 +44,15 @@ public final ServiceExceptionAssert hasType(ErrorType type) {
     public final ServiceExceptionAssert hasArgs(Arg<?>... args) {
         isNotNull();
 
-        if (!(actual.getArgs().equals(ImmutableList.copyOf(args)))) {
-            failWithMessage("Expected args to be %s, but found %s", Arrays.toString(args), actual.getArgs());
+        // toString is called on SafeArgs in SerializableError
+        List<Serializable> actualStrings = actual.getArgs().stream()
+                .map(arg -> arg instanceof SafeArg ? arg.toString() : arg)
```
Comment (meditativeape) [inline test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java]: Why only `SafeArg`? Both `SafeArg` and `UnsafeArg` share the same `toString` method.

File: test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java [hunk 2]
```diff
@@ -41,8 +44,15 @@ public final ServiceExceptionAssert hasType(ErrorType type) {
...
+                .collect(Collectors.toList());
+        List<Serializable> givenStrings = Arrays.asList(args).stream()
+                .map(arg -> arg instanceof SafeArg ? arg.toString() : arg)
+                .collect(Collectors.toList());
+        if (!(actualStrings.equals(givenStrings))) {
```
Comment (meditativeape) [inline test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java]: Should also compare the Arg names -- Arg.toString() only contains value:
```
public final String toString() {
        return String.valueOf(value);
    }
```

File: test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java (hunk 1 above)
Comment (pnepywoda) [reply test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java]: SerializableError only toString's SafeArg

**Comment by uschi2000** [comment]
Don't quite grok the rationale for this?

**Comment by pnepywoda** [comment]
CAn use Collections2.transform. Otherwise it's equals method doesn't work
vs a normal set or list to compare

**Comment by uschi2000** [comment]
I see. Then I think the correct solution is to implement here whatever assertj does when you call something like `assertThat(someList).isEqualTo(someOtherList);`. I'm pretty they're not `toString()`ing.

**Comment by pnepywoda** [comment]
The other reason is we tostring when logging. So yes I want to test it this
way anyway

**Comment by uschi2000** [comment]
I don't think you're responsible for testing the exact mechanics of the logging behavior in your project.

**Comment by pnepywoda** [comment]
It's actually "this" project not "my" project
https://github.com/palantir/http-remoting-api/blob/bb5ad9d248ce00aa275a1592fbcd11f618c38f91/errors/src/main/java/com/palantir/remoting/api/errors/SerializableError.java#L119

;)

File: test-utils/src/test/java/com/palantir/remoting/api/testing/ServiceExceptionAssertTest.java [hunk 3]
```diff
@@ -18,10 +18,12 @@
 
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
+import com.google.common.collect.Collections2;
 import com.palantir.logsafe.SafeArg;
 import com.palantir.logsafe.UnsafeArg;
 import com.palantir.remoting.api.errors.ErrorType;
 import com.palantir.remoting.api.errors.ServiceException;
+import org.assertj.core.util.Lists;
```
Comment (schlosna) [inline test-utils/src/test/java/com/palantir/remoting/api/testing/ServiceExceptionAssertTest.java]: realize this is test code, but probably want Guava's Lists

File: test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java [hunk 4]
```diff
@@ -41,8 +39,30 @@ public final ServiceExceptionAssert hasType(ErrorType type) {
...
+            }
+            if (!expected.getName().equals(actual.getName())) {
+                failWithMessage("Expected arg %s to be named %s, but found %s", i, expected.getName(),
+                        actual.getName());
+            }
+            Object expectedValue =
+                    expected.isSafeForLogging() ? expected.getValue().toString() : expected.getValue();
+            Object actualValue =
+                    actual.isSafeForLogging() ? actual.getValue().toString() : actual.getValue();
+            if (!expectedValue.equals(actualValue)) {
```
Comment (schlosna) [inline test-utils/src/main/java/com/palantir/remoting/api/testing/ServiceExceptionAssert.java]: I believe `Arg` `value` is nullable, so need to handle `.getValue().toString()` and `expectedValue.equals` may NPE

**Pull request closed without merging**